### PR TITLE
Upload audit logs to aws cloudwatch logs by vector

### PR DESCRIPTION
**Pull request by hongkailiu**
https://issues.redhat.com/browse/DPTP-746

/cc @openshift/openshift-team-developer-productivity-test-platform

### CODE REVIEW

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_fluentd-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        max_bytes 500000
+        max_lines 1000
+      </store>
+    </match>
+  output.conf: |-
+    <match kubernetes.**>
+      @type copy
+
+      <store>
+        @type google_cloud
```
Comment (stevekuznetsov) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_fluentd-configmap.yaml]: Does this work in build01 since we're on AWS?

Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_fluentd-configmap.yaml]: I have not gone that far. (Actually that is one of my questions. I do not see any info about GCE project and so on. I guess it uses the same one which the VMs are)

All i tried is
* privileged pods of a DS can run only on 3 masters
* they can access the audit-log files from the host

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      type = "file"
+      include = ["/host/var/log/kube-apiserver/audit.log"]
+      ignore_older = 86400
+
+    [sources.openshift_apiserver_audit_logs]
+      type = "file"
+      include = ["/host/var/log/openshift-apiserver/audit.log"]
+      ignore_older = 86400
+    #For debug only, will switch to https://vector.dev/docs/reference/sinks/aws_cloudwatch_logs/
+    #when debugging is done
```
Comment (hongkailiu) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml]: console is working.
Will switch to cloudwatch next week.

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      type = "file"
+      include = ["/host/var/log/kube-apiserver/audit.log"]
+      ignore_older = 86400
+
+    [sources.openshift_apiserver_audit_logs]
+      type = "file"
+      include = ["/host/var/log/openshift-apiserver/audit.log"]
+      ignore_older = 86400
+
+    [sinks.aws_cloudwatch_logs]
```
Comment (hongkailiu) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml]: TODO: need to inject auth info after James provions the IAM user.

Ref: 
https://vector.dev/docs/reference/sinks/aws_cloudwatch_logs/#aws-authentication

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
+# oc prjoect api-audit-log; oc adm policy add-scc-to-user privileged -z logging-app
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_rbac.yaml]: Can we put that into the yaml to keep this reproducible?

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      dnsPolicy: Default
+      containers:
+      - name: vector-audit-log
+        image: docker.io/timberio/vector:0.8.2-alpine
+        securityContext:
+          privileged: true
+        resources:
+          requests:
+            cpu: 500m
+            memory: 700Mi
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml]: Thats a lot, is this really needed?

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml (hunk 3 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_configmap.yaml]: Done.

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml (hunk 5 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml]: Good point.
I copied it from 

https://github.com/openshift/release/blob/a3ca62c5732160d677f08c41c8d09db1811654fe/core-services/cluster/logging/admin_fluentd-daemonset.yaml#L49-L52

Certainly api audit log shouldnt be as many as all container logs.
Let us run it sometime and get some numbers to tune.

I will leave this unsolved for now.

**Comment by hongkailiu** [comment]
Uploading is working.

https://issues.redhat.com/browse/DPTP-746?focusedCommentId=14032198&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14032198

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml (hunk 5 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml]: https://grafana-openshift-monitoring.apps.build01.ci.devcluster.openshift.com/d/85a562078cdf77779eaa1add43ccec1e/kubernetes-compute-resources-namespace-pods?orgId=1&refresh=10s&var-datasource=prometheus&var-cluster=&var-namespace=api-audit-log&var-interval=4h&from=now-24h&to=now

Screen Shot 2020-04-13 at 12 07 37 PM

Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_daemonset.yaml]: Changed to smaller numbers

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_rbac.yaml (hunk 4 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/admin_vector_rbac.yaml]: Done.

Maybe `scc/priviledged` is over-permitted.
I tried with `scc/hostaccess` and did not work out.

```bash
  Warning  FailedCreate  5s (x12 over 15s)  daemonset-controller  Error creating: pods "vector-audit-log-" is forbidden: unable to validate against any security context constraint: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed]
```

**Comment by hongkailiu** [comment]
/hold

Do not modify buildin SCCs ... define a new one.
https://coreos.slack.com/archives/GB7NB0CUC/p1586960674381500

**Comment by hongkailiu** [comment]
/hold cancel

> Do not modify buildin SCCs ... define a new one.
> https://coreos.slack.com/archives/GB7NB0CUC/p1586960674381500

Done.

**Comment by hongkailiu** [comment]
@stevekuznetsov and @alvaroaleman 

This one is ready for the next round.

File: clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+runAsUser:
+  type: RunAsAny
+seLinuxContext:
+  type: RunAsAny
+seccompProfiles:
+- '*'
+supplementalGroups:
+  type: RunAsAny
+users:
+- system:admin
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml:37]: Why system:admin?

File: clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+allowedCapabilities:
+- '*'
+allowedUnsafeSysctls:
+- '*'
+apiVersion: security.openshift.io/v1
+defaultAddCapabilities: null
+fsGroup:
+  type: RunAsAny
+groups:
+- system:cluster-admins
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml:17]: Why the groups?

File: clusters/build-clusters/01_cluster/openshift/scc/README.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# security context constraints
+
+This folder contains the SCCs that we modified.
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/scc/README.md:3]: This is wrong, the SCC that we _added_ or not?

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      terminationGracePeriodSeconds: 30
+      volumes:
+      - name: varlog
+        hostPath:
+          path: /var/log
+      - name: config-volume
+        configMap:
+          name: vector-audit-log-config
+      - name: cache-volume
+        emptyDir: {}
```
Comment (alvaroaleman) [inline clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml:66]: err...This holds the markers or not? So we definitely want this to persist across pod cycles?

File: clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml (hunk 6 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml]: Removed.

File: clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml (hunk 7 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/scc/ci-privileged_scc.yaml]: Remvoed

File: clusters/build-clusters/01_cluster/openshift/scc/README.md (hunk 8 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/scc/README.md]: Updated.

File: clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml (hunk 9 above)
Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml:66]: I understand the question.

It seems they are doing the same thing.
https://github.com/timberio/vector/blob/6b95f7e17e4af25c689a108cd3fa6941d97d5593/config/kubernetes/vector-daemonset.yaml#L91-L92

Does hostPath sound good?

Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml:66]: Lets say ... we still use `emptydir`, we could end up with uploading the same log twice.
Anything else?
They are on the masters and they should be stable.
I think `emptydir` shouldnt sound that horrible.

Comment (alvaroaleman) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml:66]: yes, hostPath would make sense here imho

Comment (alvaroaleman) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml]: I dont understand your second comment. Whats the problem with using a hostPath?

Comment (hongkailiu) [reply clusters/build-clusters/01_cluster/openshift/api_audit_log/vector-audit-log_daemonset.yaml]: Done.
Also checked on a master `/var/lib` exists.

**Comment by hongkailiu** [comment]
```
oc --context build01 get pod -n api-audit-log
NAME                     READY   STATUS    RESTARTS   AGE
vector-audit-log-4lr4g   1/1     Running   0          7m59s
vector-audit-log-9fdz4   1/1     Running   0          8m25s
vector-audit-log-gnbhn   1/1     Running   0          8m15s
```

```
### on master
sh-4.4# ls /var/lib/vector/
kube_apiserver_audit_logs  openshift_apiserver_audit_logs
```

Looks good for now.

**Pull request merged**